htlcswitch/htlcnotifier.go

```diff
@@ -0,0 +1,276 @@
+//   do not meet our outgoing link's policy (insufficient fees, for example)
+//   and when we fail to forward the payment on (insufficient outgoing
+//   capacity, or an unknown outgoing link).
+//
+// Forwarding Failure Event:
+// - Forwarding failures indicate that a htlc we forwarded has failed at
+//   another node down the route.
+// - Present for local sends and htlc forwards which fail after they left
+//   our node.
+//
+// Settle event:
+// - Settle events are present when a htlc which we added is settled through
+//   the release of a preimage.
+// - Present for local receives, and successful local sends or forwards.
+//
+// Each htlc is identified with a unique htlc key which contains its incoming
+// and outgoing channel id and index. Htlcs, and their subsequent settles or
+// fails, can be identified by these four elements. Note that sends from our
+// node will have a zero incoming channel id/index, because they originated
+// locally, and receives will have a zero outgoing channel id/index because
+// they terminate at our node.
+type HtlcNotifier struct {
+	started sync.Once
+	stopped sync.Once
+
+	ntfnServer *subscribe.Server
+}
+
+// NewHtlcNotifier creates a new HtlcNotifier which gets htlc forwarded,
+// failed and settled events from links our node has established with peers
+// and sends notifications to subscribing clients.
+func NewHtlcNotifier() *HtlcNotifier {
+	return &HtlcNotifier{
+		ntfnServer: subscribe.NewServer(),
+	}
+}
+
+// Start starts the HtlcNotifier and all goroutines it needs to consume
+// events and provide subscriptions to clients.
+func (h *HtlcNotifier) Start() error {
+	var err error
+	h.started.Do(func() {
+		log.Trace("HtlcNotifier starting")
+		err = h.ntfnServer.Start()
+	})
+	return err
+}
+
+// Stop signals the notifier for a graceful shutdown.
+func (h *HtlcNotifier) Stop() {
+	h.stopped.Do(func() {
+		if err := h.ntfnServer.Stop(); err != nil {
+			log.Warnf("error stopping htlc notifier: %v", err)
+		}
+	})
+}
+
+// SubscribeHtlcEvents returns a subscribe.Client that will receive updates
+// any time the Server is made aware of a new event.
+func (h *HtlcNotifier) SubscribeHtlcEvents() (*subscribe.Client, error) {
+	return h.ntfnServer.Subscribe()
+}
+
+// HtlcKey uniquely identifies a htlc event. It can be used to associate
+// forward events with their subsequent settle/fail event. Note that sends
+// originating from our nodes have a zero incoming channel and ID, and
+// receives terminating at our node have a zero outgoing channel and ID.
+type HtlcKey struct {
+	// IncomingChannel is the channel that the incoming htlc arrived
+	// at our node
+	// on.
+	IncomingChannel lnwire.ShortChannelID
+
+	// OutgoingChannel is the channel that the outgoing htlc left our
+	// node on.
+	OutgoingChannel lnwire.ShortChannelID
+
+	// HtlcInIndex is the index of the incoming htlc in the incoming
+	// channel.
+	HtlcInIndex uint64
```

Comment:
`IncomingHtlcIndex` or ` ChannelIn`, to make it consistent? Or just two `CircuitKey` instances.